src/Client/ClientBase.cpp

```diff
@@ -487,21 +483,21 @@ void ClientBase::receiveResult(ASTPtr parsed_query)
             /// Has the Ctrl+C been pressed and thus the query should be cancelled?
             /// If this is the case, inform the server about it and receive the remaining packets
             /// to avoid losing sync.
-            if (!query_cancelled)
+            if (!cancelled)
             {
                 auto cancel_query = [&] {
                     connection->sendCancel();
-                    query_cancelled = true;
                     if (is_interactive)
                     {
                         progress_indication.clearProgressOutput();
                         std::cout << "Cancelling query." << std::endl;
                     }
-
-                    need_cancel_query = false;
+                    cancelled = true;
+                    catch_signal.store(true);
```

Comment:
Do you mean `catch_signal.store(false)`?

Reply:
No, it will be false again after receiveResult method ends.

Reply:
But it's already true (unless the signal was caught)

Reply:
no, in handler we make it false, and it will become true again here.

Reply:
But why do we set it to true again if it was set to false in signal handler? If it is set to false, then we have received SIGINT and we should exit on the next SIGINT.

Reply:
Ok, agree.